### PR TITLE
[serve][deploy refactor][2/X] Move lightweight update logic to DeploymentVersion

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

`DeploymentVersion` already has existing support for code version + other config options that affect the version (before, only user config was taken into account). We can leverage this to do lightweight config updates, so that:
- The config's `import_path` and `runtime_env` tells us the code version for the deployments
- The logic for lightweight config updates is hidden in `DeploymentVersion`

So instead of having `DeploymentVersion` only rely on user config, we want it to rely on other info from `DeploymentInfo`, so we can decide which config options should trigger a redeployment and which shouldn't. I've added annotations `DeploymentOptionUpdateType` to fields in `DeploymentConfig` to indicate which options fall into which category as described below.

* Heavyweight options that will force replicas to restart: `version` and `ray_actor_options` (which is part of replica config)
* Lightweight options that need to call reconfigure on the replica actor: `user_config` and `graceful_shutdown_wait_loop_s`
* Lightweight options (that need to update replicas in replica state container, but won't need to call into the actual actor): `max_concurrent_queries`, `graceful_shutdown_timeout_s`, `health_check_period_s`, `health_check_timeout_s`
* Purely lightweight options: `num_replicas` and `autoscaling_config`
Screen Shot 2023-04-20 at 6 50 13 PM
Screen Shot 2023-04-20 at 6 50 31 PM

This fixes the issue of deployments not listed in config being redeployed. See the newly added test `test_deployments_not_listed_in_config`.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/version.py [hunk 1]
```diff
@@ -37,7 +65,44 @@ def __eq__(self, other: Any) -> bool:
...
+                {
+                    "max_concurrent_queries": deployment_config.max_concurrent_queries,
+                    "graceful_shutdown_timeout_s": (
+                        deployment_config.graceful_shutdown_timeout_s
+                    ),
+                    "graceful_shutdown_wait_loop_s": (
+                        deployment_config.graceful_shutdown_wait_loop_s
+                    ),
+                    "health_check_period_s": deployment_config.health_check_period_s,
+                    "health_check_timeout_s": deployment_config.health_check_timeout_s,
```
Comment (zcin) [inline python/ray/serve/_private/version.py:91]: Are these the deployment config options that, if changed, we want to redeploy the replicas?

**Comment by edoakes** [comment]
@zcin this is fantastic, it also solves a long-standing issue we've had where some of the config options are not actually dynamically updated because we weren't broadcasting the config!

File: python/ray/serve/_private/deploy_utils.py [hunk 2]
```diff
@@ -135,3 +138,14 @@ def deploy_args_to_deployment_info(
...
+
+def get_app_code_version(app_config: ServeApplicationSchema):
+    encoded = json.dumps(
+        {
+            "import_path": app_config.import_path,
+            "runtime_env": app_config.runtime_env,
+        },
+        sort_keys=True,
+    ).encode("utf-8")
+    return hashlib.md5(encoded).hexdigest()
```
Comment (edoakes) [inline python/ray/serve/_private/deploy_utils.py:151]: There's already a `ReplicaVersion` struct that implements similar logic for the `user_config` -- could we try to unify this with that?

File: python/ray/serve/_private/application_state.py [hunk 3]
```diff
@@ -182,6 +187,9 @@ def update(self):
                     self.deploy_obj_ref = None
                     logger.warning(self.app_msg)
                     return
+
+            self._process_terminating_deployments()
```
Comment (edoakes) [inline python/ray/serve/_private/application_state.py:191]: why is this moved? if the placement is important, please leave a comment indicating it

File: python/ray/serve/_private/application_state.py [hunk 4]
```diff
@@ -62,6 +62,11 @@ def delete(self):
         """Delete the application"""
         self.status = ApplicationStatus.DELETING
 
+    def update_obj_ref(self, deploy_obj_ref, deployment_timestamp):
```
Comment (edoakes) [inline python/ray/serve/_private/application_state.py:65]:
```suggestion
    def update_deploy_obj_ref(self, deploy_obj_ref, deployment_timestamp):
```

File: python/ray/serve/_private/version.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/_private/version.py:91]: It would be great if we could encode the options that require lightweight/heavyweight updates directly in the code somehow. e.g., an annotation on the `deployment_config` struct definition. can you think of a way to do that?

Comment (edoakes) [reply python/ray/serve/_private/version.py:91]: but to answer your question directly, if we are broadcasting these updates to the replicas using the existing logic for `user_config` (as I saw above), then none of these should require restarting the replicas.

**Review by edoakes — COMMENTED** [review]
@zcin the important thing to get right here is that we clearly define which fields require a "heavyweight" update (replicas restarted), encode that directly in the code, and make sure it's well tested. Can you start by listing them out explicitly in the PR description? And see my comment on if we can express it as an attribute of the config definition.

File: python/ray/serve/_private/application_state.py (hunk 3 above)
Comment (zcin) [reply python/ray/serve/_private/application_state.py]: That was a mistake, reverted

File: python/ray/serve/_private/version.py (hunk 1 above)
Comment (zcin) [reply python/ray/serve/_private/version.py]: > It would be great if we could encode the options that require lightweight/heavyweight updates directly in the code somehow. e.g., an annotation on the `deployment_config` struct definition. can you think of a way to do that?

@edoakes I've added it as an extra arg in each pydantic field

File: python/ray/serve/_private/deploy_utils.py (hunk 2 above)
Comment (zcin) [reply python/ray/serve/_private/deploy_utils.py:151]: Code version is passed to `DeploymentVersion`, which constructed in the deployment state manager after the deploy task finishes building the application, so it seems the simplest to hash the import path + runtime env into a code version when the config is first received. Maybe we can optimize this more in the future.

**Comment by zcin** [comment]
@edoakes @sihanwang41 Updated code + PR description and addressed comments, please take another look!

File: python/ray/serve/_private/deployment_state.py [hunk 5]
```diff
@@ -762,16 +769,31 @@ def start(self, deployment_info: DeploymentInfo, version: DeploymentVersion):
...
+        )
+        self._actor._graceful_shutdown_timeout_s = (
+            version.deployment_config.graceful_shutdown_timeout_s
+        )
+        self._actor._health_check_period_s = (
+            version.deployment_config.health_check_period_s
+        )
+        self._actor._health_check_timeout_s = (
+            version.deployment_config.health_check_timeout_s
         )
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_state.py:789]: we do we need to copy each of these fields individually instead of just having a copy of `deployment_config`?

Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:789]: we shouldn't be reaching into private fields if `self._actor` here

File: python/ray/serve/_private/deployment_state.py [hunk 6]
```diff
@@ -762,16 +769,31 @@ def start(self, deployment_info: DeploymentInfo, version: DeploymentVersion):
...
 
+        # Call into replica actor reconfigure() with updated config options
+        if self._version.reconfigure_actor_hash != version.reconfigure_actor_hash:
+            self._actor.reconfigure(version.deployment_config)
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_state.py:793]: I'm confused, why is this block separate from the config options that are manually set above?

File: python/ray/serve/_private/deployment_state.py [hunk 7]
```diff
@@ -1158,6 +1180,23 @@ def _set_target_state(self, target_info: DeploymentInfo) -> None:
         target_state = DeploymentTargetState.from_deployment_info(target_info)
         self._save_checkpoint_func(writeahead_checkpoints={self._name: target_state})
 
+        if self._target_state.version == target_state.version:
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_state.py:1183]: confused by this block -- shouldn't the equality check only return `True` if both code_version and config match? therefore the conditions below would never pass?

File: python/ray/serve/_private/deployment_state.py [hunk 8]
```diff
@@ -1245,44 +1288,44 @@ def _stop_wrong_version_replicas(self, max_to_stop=math.inf) -> int:
...
-            # with the correct version will be started later as part of the
-            # normal scale-up process.
-            if replica.version.code_version != self._target_state.version.code_version:
+            # If code version or ray actor options change, stop the replica. A new one
+            # with the correct version will be started later as part of the normal
+            # scale-up process.
+            if (
+                replica.version.code_version != self._target_state.version.code_version
+                or replica.version.ray_actor_options_hash
+                != self._target_state.version.ray_actor_options_hash
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_state.py:1299]: I think this logic should live in some method of the `version` instead of exposing it here, `version.requires_actor_restart(new_version)` ?

File: python/ray/serve/_private/version.py [hunk 9]
```diff
@@ -18,14 +29,16 @@ def __init__(self, code_version: Optional[str], user_config: Optional[Any] = Non
...
 
-        self.user_config = user_config
-        # TODO(simon): make this xlang compatible
-        if isinstance(user_config, bytes):
-            serialized_user_config = user_config
-        else:
-            serialized_user_config = str.encode(json.dumps(user_config, sort_keys=True))
-        self.user_config_hash = crc32(serialized_user_config)
-        self._hash = crc32(serialized_user_config + self.code_version.encode("utf-8"))
+        self.deployment_config: DeploymentConfig = deployment_config
```
Comment (edoakes) [inline python/ray/serve/_private/version.py:32]: Please add a comment for this field indicating the behavior (its options may be mutated over time so all logic that uses them should access this field directly)

File: python/ray/serve/config.py [hunk 10]
```diff
@@ -113,6 +114,19 @@ def _needs_pickle(deployment_language: DeploymentLanguage, is_cross_language: bo
         return False
 
 
+class DeploymentOptionUpdateType(str, Enum):
+    # Nothing needs to be done other than setting the target state.
+    LightWeight = "LightWeight"
```
Comment (edoakes) [inline python/ray/serve/config.py:119]: let's name this something more descriptive. what options actually fit into this btw? is it basically `num_replicas` and so forth?

File: python/ray/serve/config.py [hunk 11]
```diff
@@ -141,21 +155,37 @@ class DeploymentConfig(BaseModel):
...
-        DEFAULT_GRACEFUL_SHUTDOWN_TIMEOUT_S  # noqa: E501
+    graceful_shutdown_timeout_s: NonNegativeFloat = Field(
+        default=DEFAULT_GRACEFUL_SHUTDOWN_TIMEOUT_S,
+        update_type=DeploymentOptionUpdateType.NeedsReconfigure,
     )
-    graceful_shutdown_wait_loop_s: NonNegativeFloat = (
-        DEFAULT_GRACEFUL_SHUTDOWN_WAIT_LOOP_S  # noqa: E501
+    graceful_shutdown_wait_loop_s: NonNegativeFloat = Field(
+        default=DEFAULT_GRACEFUL_SHUTDOWN_WAIT_LOOP_S,
+        update_type=DeploymentOptionUpdateType.NeedsActorReconfigure,
```
Comment (edoakes) [inline python/ray/serve/config.py:174]: Thanks for adding this, it is a big improvement!

**Review by edoakes — COMMENTED** [review]
Huge improvement, I think it's almost there!

File: python/ray/serve/_private/deploy_utils.py [hunk 12]
```diff
@@ -135,3 +138,14 @@ def deploy_args_to_deployment_info(
...
     )
+
+
+def get_app_code_version(app_config: ServeApplicationSchema):
+    encoded = json.dumps(
+        {
+            "import_path": app_config.import_path,
+            "runtime_env": app_config.runtime_env,
+        },
+        sort_keys=True,
```
Comment (sihanwang41) [inline python/ray/serve/_private/deploy_utils.py:149]: nice with sort_keys.

File: python/ray/serve/_private/deployment_state.py [hunk 13]
```diff
@@ -1180,10 +1219,14 @@ def deploy(self, deployment_info: DeploymentInfo) -> bool:
         existing_info = self._target_state.info
         if existing_info is not None:
             # Redeploying should not reset the deployment's start time.
-            deployment_info.start_time_ms = existing_info.start_time_ms
+            if not self._target_state.deleting:
```
Comment (sihanwang41) [inline python/ray/serve/_private/deployment_state.py:1222]: `if not self._target_state.deleting`, what the reason for adding this check? is this an existing bug?

File: python/ray/serve/_private/replica.py [hunk 14]
```diff
@@ -484,26 +479,36 @@ async def invoke_single(self, request_item: Query) -> Tuple[Any, bool]:
...
         return result, success
 
-    async def reconfigure(self, user_config: Any):
+    async def reconfigure(self, deployment_config: DeploymentConfig):
         async with self.rwlock.writer_lock:
-            self.user_config = user_config
-            self.version = DeploymentVersion(
-                self.version.code_version, user_config=user_config
+            user_config_changed = False
+            if (
```
Comment (sihanwang41) [inline python/ray/serve/_private/replica.py:485]: nit: should we directly `if self.deployment_config.user_config != deployment_config.user_config`

File: python/ray/serve/_private/version.py [hunk 15]
```diff
@@ -18,14 +29,16 @@ def __init__(self, code_version: Optional[str], user_config: Optional[Any] = Non
...
+        self.ray_actor_options: Dict = ray_actor_options
+        self.compute_hashes()
```
Comment (sihanwang41) [inline python/ray/serve/_private/version.py:34]: I assume we don't allow anyone to update the attribute, since it will change the hash value. I suggest to change the class DeploymentVersion into pydantic and freeze it after initialization. Or we raise error when anyone update the attribute.

File: python/ray/serve/config.py (hunk 11 above)
Comment (sihanwang41) [reply python/ray/serve/config.py:174]: I am into it, so clear.

File: python/ray/serve/config.py (hunk 10 above)
Comment (zcin) [reply python/ray/serve/config.py:119]: It's `num_replicas` and `autoscaling_config`.

File: python/ray/serve/_private/deployment_state.py (hunk 7 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_state.py:1183]: This is what I use for `__hash__` in `DeploymentVersion`:
image

So basically `num_replicas` and `autoscaling_config` (the only two lightweight options) are excluded when calculating version.

File: python/ray/serve/_private/deployment_state.py (hunk 6 above)
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:793]: Ah, I understand now, thanks 👍

File: python/ray/serve/_private/deployment_state.py (hunk 7 above)
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:1183]: Ah, I get it now. I find this pretty confusing -- intuitively I'd expect the equality check to check _all_ options and then maybe we'd have a separate method for checking partial equality like this

File: python/ray/serve/tests/test_config_files/pid.py [hunk 16]
```diff
@@ -21,7 +21,7 @@ def __init__(self, dag: RayServeDAGHandle):
         self.dag = dag
 
     async def __call__(self):
-        return await self.dag.remote()
+        return await (await self.dag.remote())
```
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/tests/test_config_files/pid.py:24]: I'm pretty sure this code didn't work before 🤦‍♀️. Not sure how all the tests were passing in `test_deploy_config_update`, but some of them were because I wasn't casting the response to int, so I've updated the test to do that.

File: python/ray/serve/_private/replica.py (hunk 14 above)
Comment (zcin) [reply python/ray/serve/_private/replica.py:485]: When the replica is first initialized, deployment_config is None.

File: python/ray/serve/_private/version.py (hunk 15 above)
Comment (zcin) [reply python/ray/serve/_private/version.py:34]: Would it become too heavyweight if we change it to pydantic? I'm in favor of making deployment version immutable, though.

File: python/ray/serve/_private/deployment_state.py (hunk 7 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_state.py:1183]: Hm yeah I can see how that might be confusing ... 
I think version is used specifically for detecting "old replicas with outdated versions" that need updating (see [code pointer](https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/deployment_state.py#L1240)). So when the target state is updated with new `num_replicas` and `autoscaling_config` values, that is automatically picked up by deployment state manager, but has nothing to do with updating old replicas.

Perhaps a better way to deal with this is rename DeploymentVersion to ReplicaVersion lol. Maybe we can also move options like `max_concurrent_queries` into `ReplicaConfig` instead of `DeploymentConfig`.

File: python/ray/serve/_private/version.py (hunk 9 above)
Comment (zcin) [reply python/ray/serve/_private/version.py:34]: Added!

File: python/ray/serve/_private/deployment_state.py (hunk 5 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_state.py]: Good point, I've simplified things a bit!

File: python/ray/serve/_private/deployment_state.py (hunk 7 above)
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:1168]: that makes a lot of sense to me... `ReplicaConfig` is a subset of `DeploymentConfig` and contains all of the things that get passed to each replica

if it's a big change do it in a follow up?

**Review by edoakes — COMMENTED** [review]
@zcin can you add an integration test that `graceful_shutdown_period_s` is correctly updated without restarting the replica? Ideally we'd have this for each of the options that can be set.

You can do this by:

- Start with an extremely long `graceful_shutdown_period_s`.
- Update it to be very short.
- Check that the replica ID/actor ID doesn't change.
- Send a request that hangs forever.
- Shutdown or do an update, check that it completes quickly.

File: python/ray/serve/_private/version.py (hunk 15 above)
Comment (sihanwang41) [reply python/ray/serve/_private/version.py:36]: both sounds good to me, "freezing the class" makes the code less error-proned.

File: python/ray/serve/_private/deployment_state.py (hunk 13 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_state.py:1207]: I'm not sure, I don't know how often this case will occur (probably not often?). 
This is specifically for when a deployment has been deleted (e.g. the deployment graph changed) and then redeployed (e.g. the deployment graph changed back). 

The biggest change I made here is on line 1211, so that if all configs are the same but the current target state is deleting, the deployment isn't blocked from redeploying. I ran into this problem when debugging some unit tests, not sure why this pops up in my PR specifically.

File: python/ray/serve/_private/version.py (hunk 15 above)
Comment (zcin) [reply python/ray/serve/_private/version.py:36]: Yeah, agreed. I'll probably look into this in a potential follow up PR.

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
@zcin serve tests failing.

File: python/ray/serve/_private/deployment_state.py [hunk 17]
```diff
@@ -1243,48 +1271,47 @@ def _stop_wrong_version_replicas(self, max_to_stop=math.inf) -> int:
...
         )
-        replicas_stopped = False
+        replicas_changed = False
         code_version_changes = 0
-        user_config_changes = 0
+        reconfigure_changes = 0
         for replica in replicas_to_update:
-            # If the code version is a mismatch, we stop the replica. A new one
-            # with the correct version will be started later as part of the
+            replicas_changed = True
```
Comment (sihanwang41) [inline python/ray/serve/_private/deployment_state.py:1278]: will this trigger `_stop_wrong_version_replicas` returning True and potentially trigger a long poll broadcast which is not necessary?

File: python/ray/serve/_private/deployment_state.py [hunk 18]
```diff
@@ -1243,48 +1271,47 @@ def _stop_wrong_version_replicas(self, max_to_stop=math.inf) -> int:
             max_replicas=max_to_stop,
             ranking_function=rank_replicas_for_stopping,
         )
-        replicas_stopped = False
+        replicas_changed = False
```
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/deployment_state.py:1277]: Renamed to `_stop_or_update_outdated_version_replicas`

File: python/ray/serve/_private/deployment_state.py (hunk 17 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_state.py]: That's a good point, it's only necessary for replicas that have max concurrent queries updated. I've changed the code to only return True in that case.

**Comment by sihanwang41** [comment]
could you rebase master? dashboard tests are fixed.

**Review by sihanwang41 — APPROVED** [review]

**Comment by zcin** [comment]
@edoakes Tests look good, dataset tests unrelated and failing on master.
